### PR TITLE
Ensure Kafka Sinks with keys only write to Kafka from one worker

### LINKED ISSUE #6538: Kafka Sink with ENVELOPE UPSERT should not write to consistency topic from multiple workers
Most parts of the sink machinery currently work on the assumption that only one sink instance will write to both the output topic and the consistency topic (when enabled). The main reason for this is that the consistency topic basically keeps track of what times are valid and how many messages are in a given time interval. When multiple workers write to this topic it can get scrambled. Additionally, it currently would not work to synchronize transactional writes across multiple active sink instances for one topic.

We try to enforce the requirement here:
https://github.com/MaterializeInc/materialize/blob/7678c75b718b27fd70cbdef17ec66bedd700c41b/src/dataflow/src/render/sinks.rs#L84

However, this doesn't work because the later `arrange_by_key()` shuffles the data again (by key):
https://github.com/MaterializeInc/materialize/blob/7678c75b718b27fd70cbdef17ec66bedd700c41b/src/dataflow/src/render/sinks.rs#L111

This is where the Kafka sink assumes the requirement which is not true: https://github.com/MaterializeInc/materialize/blob/7678c75b718b27fd70cbdef17ec66bedd700c41b/src/dataflow/src/sink/kafka.rs#L564

I think the solution is for the sink to enforce this requirement itself instead of relying on far-away code to uphold this. The responsibilities/assumptions are quite far spread in this case. Also, we should add a test for this assumption.

### PR DESCRIPTION
**Pull request by aljoscha**
This fixes #6538.

Please see the commit descriptions for details.

### CODE REVIEW

**Comment by aljoscha** [comment]
This might still fail for some tests that don't yet work with `count(workers) > 1`.

File: test/testdrive/mzcompose.yml [hunk 1]
```diff
@@ -98,7 +98,7 @@ services:
     mzbuild: materialized
     command: >-
       --data-directory=/share/mzdata
-      -w1 --experimental
+      -w5 --experimental
```
Comment (philip-stoev) [inline test/testdrive/mzcompose.yml:101]: Thanks a lot!

**Review by philip-stoev — APPROVED** [review]

**Comment by ruchirK** [comment]
I'll look at this in an hour or so (going to commute to the office today)

**Comment by aljoscha** [comment]
By the way, I made the encoding operator "cooperative" (by using fuel and yielding early) because otherwise scheduling durations can get a bit high. When the operator is greedy high scheduling durations are between 4s and 8s. With 10k fuel, scheduling durations are between 1s and 2s.

Also, right now this is only tested by increasing the number of workers that we use for testdrive tests. I could add a dedicated test setup, similar to what we have in `kafka-exactly-once` where we control the complete test setup.

**Comment by ruchirK** [comment]
sorry I totally dropped the ball on this looking now!

File: src/dataflow/src/sink/kafka.rs [hunk 2]
```diff
@@ -574,3 +581,161 @@ where
 
     Box::new(KafkaSinkToken { shutdown_flag })
 }
+
+/// An operator for encoding Rows as Avro that can yield, and doesn't have to process all its in
```
Comment (ruchirK) [inline src/dataflow/src/sink/kafka.rs:585]: nit: input is one word

File: src/dataflow/src/sink/kafka.rs [hunk 3]
```diff
@@ -574,3 +581,161 @@ where
...
+
+            if records.is_empty() {
+                // drop our capability for this time
+                stash.remove(&lowest_ts);
+            }
+        }
+
+        if !stash.is_empty() {
+            activator.activate();
+            return true;
```
Comment (ruchirK) [inline src/dataflow/src/sink/kafka.rs:665]: nit: I don't think you need the return statement for correctness

File: src/dataflow/src/sink/kafka.rs [hunk 4]
```diff
@@ -574,3 +581,161 @@ where
...
+    value_schema_id: i32,
+    fuel: usize,
+    sink_name: String,
+) -> Stream<G, ((Option<Vec<u8>>, Option<Vec<u8>>), u64, isize)>
+where
+    G: Scope<Timestamp = Timestamp>,
+{
+    let name = format!("{}-avro_encode", sink_name);
+
+    let mut builder = OperatorBuilder::new(name, input_stream.scope());
```
Comment (ruchirK) [inline src/dataflow/src/sink/kafka.rs:600]: so im still working through this change but this seems like a really low level operator building api. is there any reason we can't use `unary_frontier` or `unary` here (I see that we use `OperatorBuilder` elsewhere as well). do we need it to use `build_reschedule` is that the reason?

**Review by ruchirK — COMMENTED** [review]
I've taken a quick preliminary look. I need to step away for something rn but will take a closer look in an hour. so far everything looks reasonable!

File: src/dataflow/src/sink/kafka.rs [hunk 5]
```diff
@@ -574,3 +581,161 @@ where
...
+            let lowest_ts = stash
+                .keys()
+                .min_by(|x, y| x.time().cmp(y.time()))
+                .expect("known to exist")
+                .clone();
+            let records = stash.get_mut(&lowest_ts).expect("known to exist");
+
+            let mut session = output.session(&lowest_ts);
+            let num_records_to_drain = cmp::min(records.len(), fuel_remaining);
+            // let num_records_to_drain = records.len();
```
[the remark this answers is not in the document]
Comment (aljoscha) [reply src/dataflow/src/sink/kafka.rs:644]: sorry, leftover 👌

File: src/dataflow/src/sink/kafka.rs (hunk 3 above)
Comment (aljoscha) [reply src/dataflow/src/sink/kafka.rs:665]: Yes, I left it in more for clarity, but I can remove. 👍

File: src/dataflow/src/sink/kafka.rs (hunk 4 above)
Comment (aljoscha) [reply src/dataflow/src/sink/kafka.rs:600]: Yes, it is because the operator can yield while still having records in the stash. I thought we then need the more low-level API that allows an operator to signal that it would like to run in the future by a) returning `true` and, b) re-scheduling using the `Activator`.

File: src/dataflow/src/sink/kafka.rs [hunk 6]
```diff
@@ -26,10 +27,14 @@ use rdkafka::error::{KafkaError, RDKafkaErrorCode};
 use rdkafka::message::Message;
 use rdkafka::producer::Producer;
 use rdkafka::producer::{BaseRecord, DeliveryResult, ProducerContext, ThreadedProducer};
+use timely::dataflow::channels::pact::Exchange as ExchangePact;
```
Comment (ruchirK) [inline src/dataflow/src/sink/kafka.rs:30]: I don't think you need to rename this import

```
diff --git a/src/dataflow/src/sink/kafka.rs b/src/dataflow/src/sink/kafka.rs
index 1629feb0a..6681f8610 100644
--- a/src/dataflow/src/sink/kafka.rs
+++ b/src/dataflow/src/sink/kafka.rs
@@ -27,8 +27,7 @@ use rdkafka::error::{KafkaError, RDKafkaErrorCode};
 use rdkafka::message::Message;
 use rdkafka::producer::Producer;
 use rdkafka::producer::{BaseRecord, DeliveryResult, ProducerContext, ThreadedProducer};
-use timely::dataflow::channels::pact::Exchange as ExchangePact;
-use timely::dataflow::channels::pact::Pipeline;
+use timely::dataflow::channels::pact::{Exchange, Pipeline};
 use timely::dataflow::operators::generic::builder_rc::OperatorBuilder;
 use timely::dataflow::operators::generic::{FrontieredInputHandle, OutputHandle};
 use timely::dataflow::operators::Capability;
@@ -692,7 +691,7 @@ where
     let name = format!("{}-sort_on_one_worker", sink_name);

     let sorted_output_stream = input_stream.unary_frontier(
-        ExchangePact::new(move |_| sink_id.hashed()),
+        Exchange::new(move |_| sink_id.hashed()),
         &name,
         move |_, _| {
             let mut stash = HashMap::new();
```
compiles for me

File: src/dataflow/src/sink/kafka.rs [hunk 7]
```diff
@@ -285,6 +289,21 @@ where
...
+        stream,
+        encoder,
+        key_schema_id,
+        value_schema_id,
+        connector.fuel,
+        name.clone(),
+    );
+
+    let stream = sort_on_one_worker(&stream, name.clone(), id);
+
```
Comment (ruchirK) [inline src/dataflow/src/sink/kafka.rs:306]: I know this wasn't explicitly part of your change but could we get a later commit that renames `s` on line 310 to something else `sink_state` maybe? as is `s` is hard to follow through the code

File: src/dataflow/src/sink/kafka.rs (hunk 3 above)
Comment (ruchirK) [reply src/dataflow/src/sink/kafka.rs:665]: ok so in the other case where we use `build_reschedule` in `sink_logic` in this file, whenever we return true to `build_reschedule` we also manually specify when we want to be activated
I'm looking at the following:
```
552         if !ready_rows.is_empty() {
553             // We need timely to reschedule this operator as we have pending
554             // items that we need to send to Kafka
555             s.activator.activate();
556             return true;
557         }
558
559         if in_flight > 0 {
560             // We still have messages that need to be flushed out to Kafka
561             // Let's make sure to keep the sink operator around until
562             // we flush them out
563             s.activator.activate_after(Duration::from_secs(5));
564             return true;
565         }
566
```

my understanding is that a return value of `true` indicates to timely "you do not have permission to delete this operator", but it doesn't actually reschedule the operator and the operator needs to use the activator to do so manually

Comment (ruchirK) [reply src/dataflow/src/sink/kafka.rs:665]: oh no hang on -- you can just do this

```
@@ -664,8 +663,8 @@ where
             activator.activate();
             return true;
         }
-        // signal that we're still incomplete
-        !stash.is_empty()
+        // signal that we're complete.
+        false
     };
     ```
  which i think is much simpler

**Comment by ruchirK** [comment]
I haven't looked at the test changes super closely but do we fail the new tests if we run without the corresponding changes to the kafka sink?

File: src/dataflow/src/sink/kafka.rs [hunk 8]
```diff
@@ -574,3 +581,161 @@ where
...
+    let activator = input_stream
+        .scope()
+        .activator_for(&builder.operator_info().address[..]);
+
+    let mut stash: HashMap<Capability<Timestamp>, Vec<_>> = HashMap::new();
+    let mut vector = Vec::new();
+    let mut encode_logic = move |input: &mut FrontieredInputHandle<
+        Timestamp,
+        ((Option<Row>, Option<Row>), Timestamp, Diff),
```
Comment (ruchirK) [inline src/dataflow/src/sink/kafka.rs:613]: I don't know if you can do anything about this but the formatting here is hard to read :/

File: src/dataflow/src/sink/kafka.rs [hunk 9]
```diff
@@ -574,3 +581,161 @@ where
 
     Box::new(KafkaSinkToken { shutdown_flag })
 }
+
+/// An operator for encoding Rows as Avro that can yield, and doesn't have to process all its in
+/// put in one invocation.
+fn avro_encode_stream<G>(
```
Comment (frankmcsherry) [inline src/dataflow/src/sink/kafka.rs:587]: I'm going to be a pain in the butt and ask for some more documentation here. Can you say what properties the input stream should have coming in, and what properties the output stream is guaranteed to have at its conclusion? There are nearby bits of code that want to assume things like operators don't shuffle dat, maybe maintain partitioning by key. This code may not have any thoughts there, or it might, but if you can jot them down as part of the method doccomments, that would avoid future confusion.

**Comment by ruchirK** [comment]
Everything else looks ok to me -- feel free to merge or ping if there's other stuff you want me to look at!

File: src/dataflow/src/sink/kafka.rs [hunk 10]
```diff
@@ -574,3 +581,161 @@ where
...
+    output_stream
+}
+
+/// Pulls all data to one worker by hashing on the sink id. Sorts and emits data
+/// that is before the frontier.
+fn sort_on_one_worker<G>(
+    input_stream: &Stream<G, ((Option<Vec<u8>>, Option<Vec<u8>>), u64, isize)>,
+    sink_name: String,
+    sink_id: GlobalId,
+) -> Stream<G, ((Option<Vec<u8>>, Option<Vec<u8>>), u64, isize)>
```
Comment (frankmcsherry) [inline src/dataflow/src/sink/kafka.rs:688]: This seems like a weird timely dataflow operator, from personal experience. It could just be evolution from our code base, but: generally, each operator has the ability to drive its own data distribution rules, and if it needs its input to be e.g. on one worker, it should do that (to make sure that the contract is maintained). 

If it turns out that the problem is that too many operators need this behavior, we can try and wrap it up. For example, timely's `Notificator` struct wraps input streams and provide their contents in time order.

Comment (frankmcsherry) [reply src/dataflow/src/sink/kafka.rs:688]: Like above, this would be a great place to put more doccomments about what the method assumes about its input, and what it guarantees about its output. For example, what are the data sorted by, and how should a user use the output of this method (I would personally put big red warning flags that the ordering is potentially destroyed if a user invokes an operator that exchanges the data, which is one of the reasons I am a bit weirded out about it).

File: src/dataflow/src/sink/kafka.rs (hunk 3 above)
Comment (aljoscha) [reply src/dataflow/src/sink/kafka.rs:665]: thanks! 👌

File: src/dataflow/src/sink/kafka.rs (hunk 6 above)
Comment (aljoscha) [reply src/dataflow/src/sink/kafka.rs:30]: Ah yes, this was from another version where I also had also imported the `Exchange` operator (https://github.com/TimelyDataflow/timely-dataflow/blob/221f98930c92b75d215c720e88f559cdab7d9147/timely/src/dataflow/operators/exchange.rs#L9)

File: src/dataflow/src/sink/kafka.rs (hunk 9 above)
Comment (aljoscha) [reply src/dataflow/src/sink/kafka.rs:587]: Absolutely!

File: src/dataflow/src/sink/kafka.rs (hunk 8 above)
Comment (aljoscha) [reply src/dataflow/src/sink/kafka.rs:613]: Very strange, yes. But this is what rustfmt will give me. 😞

File: src/dataflow/src/sink/kafka.rs (hunk 10 above)
Comment (aljoscha) [reply src/dataflow/src/sink/kafka.rs:688]: Agreed! What got me started on this is that the Kafka sink operator was assuming that `render/sinks.rs` would a) sort the data by timestamp, and b) exchange it all to one operator. This assumption is incorrect so I moved that responsibility "closer" to the Kafka operator.

I think I didn't move it close enough and should include the "exchanging all to one operator and sort it " part into the Kafka sink operator (via its Pact and then by sorting internally).

I will do that now, but I also need to think about future Kafka Sinks that don't need this sorting/exchanging to one operator. Thinking about a CDCv2 sink here. It might turn out that the existing Kafka Sink and that one can't share much because their requirements are too different.

**Comment by aljoscha** [comment]
I updated the code to remove the extra sorting operator. Turns out we don't need to sort for what the sink currently does, we still write data in timestamp order in the Kafka sink, though. I also factored out the "write to kafka" part and added docs that describe the intended behaviour/guarantees/requirements:
 - https://github.com/MaterializeInc/materialize/pull/6574/files#diff-c58795b3c30d976cf371ccdc9773f63e648255ca93a6bf17af92ead8773540e4R272
 -  https://github.com/MaterializeInc/materialize/pull/6574/files#diff-c58795b3c30d976cf371ccdc9773f63e648255ca93a6bf17af92ead8773540e4R611

Please take another look.

File: src/dataflow/src/sink/kafka.rs [hunk 11]
```diff
@@ -242,7 +244,52 @@ pub fn kafka<G>(
...
+///
+/// This operator exchanges all updates to a single worker by hashing on the given sink `id`.
+///
+/// Updates are only sent to Kafka once the input frontier has passed their `time`. Updates are
+/// sent in ascending timestamp order. The order of updates at the same timstamp will not be changed.
+/// However, it is important to keep in mind that this operator exchanges updates so if the input
+/// stream is sharded updates will likely arrive at this operator in some non-deterministic order.
+///
+/// Updates that are not beyond the given [`SinkAsOf`] and/or the `gate_ts` in
+/// [`KafkaSinkConnector`] will be discarded without producing them.
```
Comment (frankmcsherry) [inline src/dataflow/src/sink/kafka.rs:282]: This is a slightly different use of `as_of` than we use elsewhere in the code base. Idk if it is new here or something being borrowed, but: the conventional use of `as_of` is to indicate that all times should be advanced to the frontier described by `as_of`. Whether such updates should be discarded or not (vs advanced) is traditionally a bool `snapshot`.

It is possible that this isn't a new issue in sink code; if it isn't new we should file an issue to straighten out the language.

File: src/dataflow/src/sink/kafka.rs [hunk 12]
```diff
@@ -574,3 +607,133 @@ where
...
+
+/// Encodes a stream of `(Option<Row>, Option<Row>)` updates using Avro.
+///
+/// This operator will only encode `fuel` number of updates per invocation. If necessary, it will
+/// stash updates and use an [`timely::scheduling::Activator`] to re-schedule future invocations.
+///
+/// Input [`Row`] updates must me compatible with the given [`Encoder`].
+///
+/// Updates that are not beyond the given [`SinkAsOf`] and/or the `gate_ts` will be discarded
+/// without encoding them.
```
Comment (frankmcsherry) [inline src/dataflow/src/sink/kafka.rs:619]: See comment above about use of `as_of`. Same concern applies here.

File: src/dataflow/src/sink/kafka.rs [hunk 13]
```diff
@@ -574,3 +607,133 @@ where
...
+) -> Stream<G, ((Option<Vec<u8>>, Option<Vec<u8>>), Timestamp, Diff)>
+where
+    G: Scope<Timestamp = Timestamp>,
+{
+    let name = format!("{}-avro_encode", name_prefix);
+
+    let mut builder = OperatorBuilder::new(name, input_stream.scope());
+    let mut input = builder.new_input(&input_stream, Pipeline);
+    let (mut output, output_stream) = builder.new_output();
+    builder.set_notify(false);
```
Comment (frankmcsherry) [inline src/dataflow/src/sink/kafka.rs:646]: I'm not certain if this is intended, but it communicates to timely that this operator does not need to be notified about changes in its input frontier. It is a bit odd to have this and also a `FrontieredInputHandle`, which suggests that you are interested in the frontier of the input handle. Easiest fix is to remove this line and then chat about the implications of this call. It's not clearly wrong, in that I don't see an `input.frontier` use, but worth discussing!

File: src/dataflow/src/sink/kafka.rs [hunk 14]
```diff
@@ -562,9 +596,8 @@ where
     };
 
     // We want exactly one worker to send all the data to the sink topic.
-    // This should already have been handled upstream (in render/mod.rs),
-    // so we can use `Pipeline` here.
-    let mut input = builder.new_input(stream, Pipeline);
+    let mut input = builder.new_input(&stream, Exchange::new(move |_| id.hashed()));
```
Comment (frankmcsherry) [inline src/dataflow/src/sink/kafka.rs:599]: I believe this will re-evaluate `id.hashed()` for each record. Probably best to compute once outside the closure and then jam the result in here.

**Review by frankmcsherry — APPROVED** [review]
The structure of everything looks good to me. I left some comments that .. probably should be sorted out but are none of them fatal. Thanks very much!

File: src/dataflow/src/sink/kafka.rs (hunk 13 above)
Comment (aljoscha) [reply src/dataflow/src/sink/kafka.rs:646]: Oh boy, sorry about that one! It's left over from an in-progress version where I added metrics for tracking the frontier. I'll change it to `InputHandle` but leave in `set_notify(false)` because we're not interested in frontier updates. The actual Kafka Producer operator has this on `true` (the default) because it does care about the frontier.

File: src/dataflow/src/sink/kafka.rs (hunk 14 above)
Comment (aljoscha) [reply src/dataflow/src/sink/kafka.rs:599]: 👍

File: src/dataflow/src/sink/kafka.rs (hunk 13 above)
Comment (aljoscha) [reply src/dataflow/src/sink/kafka.rs:647]: Oh boy, sorry about that one! It's left over from an in-progress version where I added metrics for tracking the frontier. I'll change it to `InputHandle` but leave in `set_notify(false)` because we're not interested in frontier updates. The actual Kafka Producer operator has this on `true` (the default) because it does care about the frontier.

File: src/dataflow/src/sink/kafka.rs (hunk 11 above)
Comment (aljoscha) [reply src/dataflow/src/sink/kafka.rs:282]: It's a new-ish thing, added right before my time: https://github.com/MaterializeInc/materialize/commit/06c0f76257d7582793540906c83a89e681d5194d
